### DISCUSSION TITLE
Google oauth2 for specific domains

### DISCUSSION THREAD

**Discussion by diegocgaona** [fastapi-users/fastapi-users#1109] [Q&A] [answered]
I want to make logins with Google Sign-In, but only for a specific domain.

I saw that you can use the URI parameter `hd` (https://developers.google.com/identity/openid-connect/openid-connect#hd-param) to google optimize the UI for an domain and then validate (https://developers.google.com/identity/openid-connect/openid-connect#validatinganidtoken) the returned ID token `hd` (https://developers.google.com/identity/openid-connect/openid-connect#obtainuserinfo) field to check if it is my desired domain.

It's possible to do it with FastAPI Users?

**Comment by frankie567** [comment, marked as answer]
Two things here:

1. Customizing the Google UI is done through specific query parameters to pass to the authorize URL. Currently, it's not possible to add arbitrary parameters to the authorize URL. This could be an interesting feature, I'll add it to the backlog. In the meantime, you'll need to re-implement the `/authorize` route:

https://github.com/fastapi-users/fastapi-users/blob/baca17d9e029aafb757e128cedbb6782e45bcd89/fastapi_users/router/oauth.py#L53-L74

2. Checking for some constraints after callback. This can be done by overloading the `callback` method of `UserManager`. Something like this:

```py
from typing import Optional

import httpx
from fastapi import HTTPException, Request
from fastapi_users import BaseUserManager, UUIDIDMixin


class UserManager(UUIDIDMixin, BaseUserManager[User, uuid.UUID]):
    # ...
    async def oauth_callback(
        self,
        oauth_name: str,
        access_token: str,
        account_id: str,
        account_email: str,
        expires_at: Optional[int] = None,
        refresh_token: Optional[str] = None,
        request: Optional[Request] = None,
        *,
        associate_by_email: bool = False
    ) -> User:
        # You have here a valid access token to Google API,
        # so you can call it to get the data you need
        async with httpx.AsyncClient() as client:
            response = await client.get(
                "https://openidconnect.googleapis.com/v1/userinfo",
                headers={"Authorization": f"Bearer {access_token}"}
            )
            data = response.json()
            if data["hd"] != "domain.com":
                raise HTTPException(detail="Invalid domain")

        return await super().oauth_callback(
            oauth_name,
            access_token,
            account_id,
            account_email,
            expires_at=expires_at,
            refresh_token=refresh_token,
            request=request,
            associate_by_email=associate_by_email,
        )
```
